users/user-guide-query/multi-stage-query/operator-types/aggregate.md

```diff
@@ -0,0 +1,194 @@
+---
+description: >-
+  Describes the aggregate relation operator in the multi-stage query engine.
+---
+
+# Aggregate relational operator
+
+The aggregate operator is used to perform calculations on a set of rows and return a single row of results.
+
+This page describes the aggregate operator defined in the relational algebra used by multi-stage queries.
+This operator is generated by the multi-stage query engine when you use _aggregate functions_ in a query either with or
+without a `group by` clause.
+
+## Implementation details
+Aggregate operations may be expensive in terms of memory, CPU and network usage. 
+As explained in [understanding stages](../understanding-stages.md), the multi-stage query engine breaks down the query 
+into multiple stages and each stage is then executed in parallel on different _workers_.
+Each worker processes a subset of the data and sends the results to the _coordinator_ which then aggregates the results.
+When possible, the multi-stage query engine will try to apply a divide-and-conquer strategy to reduce the amount of data
+that needs to be processed in the _coordinator_ stage.
+
+For example if the aggregation function is a sum, the engine will try to sum the results of each worker before sending
+the partial result to the _coordinator_, which would then sum the partial results in order to get the final result.
+But some aggregation functions, like `count(distinct)`, cannot be computed in this way and require all the data to be
+processed in the _coordinator_ stage.
+
+In Apache Pinot 1.1.0, the multi-stage query engine always keeps the data in memory.
+This means that the amount of memory used by the engine is proportional to the number of groups generated by the
+`group by` clause and the amount of data that needs to be kept for each group (which depends on the aggregation 
+function).
+
+Even when the aggregation function is a simple `count`, which only requires to keep a long for each group in memory,
+the amount of memory used can be high if the number of groups is high.
+This is why the engine limits the number of groups.
+By default, this limit is 100.000, but this can be changed by providing hints.
+
+### Blocking nature
+The aggregate operator is a blocking operator. It needs to consume all the input data before emitting the result.
+
+## Hints
+
+### num_groups_limit
+Type: Integer
+
+Default: 100.000
+
+Defines the max number of groups that can be created by the `group by` clause. 
+If the number of groups exceeds this limit, the query will not fail but will stop the execution.
```

Comment:
another clarification on the implication of this:
if num groups is exceeded, will we continue reading the rest of the data to look for groups we already have? i assume we'll finish reading through all data, but not pick any new groups?

Reply:
That is what I would expect, yes. But I reading the code it looks to me that we do not enforce that but instead continue working as normal and at the end we just mark the flag in the stats. I may be wrong, I wasn't working on that part of the code and I didn't dedicate much time reading all possible implementations (there are at least 3 different classes involved when new blocks are received)